### PR TITLE
Fixed #25296 -- Prevented related object cache pollution when create() fails due to an unsaved object.

### PR DESCRIPTION
**Pull request by raphaelmerx**
https://code.djangoproject.com/ticket/25296

See behavior in `one_to_one.tests.OneToOneTests.test_unsaved_object`: because `Restaurant.objects.create` failed, `place._restaurant_cache` should not exist.

Any feedback is much appreciated!

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: tests/one_to_one/tests.py [hunk 1]
```diff
@@ -135,9 +135,11 @@ def test_unsaved_object(self):
         should raise an exception.
         """
         place = Place(name='User', address='London')
+        self.assertRaises(Restaurant.DoesNotExist, lambda: place.restaurant)
```
Comment (aaugustin) [inline tests/one_to_one/tests.py]: Please use this style:

    with self.assertRaises(Restaurant.DoesNotExist):
        place.restaurant

File: django/db/models/base.py [hunk 2]
```diff
@@ -643,6 +643,8 @@ def save(self, force_insert=False, force_update=False, using=None,
                 # constraints aren't supported by the database, there's the
                 # unavoidable risk of data corruption.
                 if obj and obj.pk is None:
+                    if not field.related.multiple:
```
Comment (timgraham) [inline django/db/models/base.py]: As noted on the build failure, "Usage of field.related has been deprecated. Use field.remote_field instead."

**Comment by timgraham** [comment]
buildbot, test this please.

File: django/db/models/base.py [hunk 3]
```diff
@@ -643,6 +643,8 @@ def save(self, force_insert=False, force_update=False, using=None,
                 # constraints aren't supported by the database, there's the
                 # unavoidable risk of data corruption.
                 if obj and obj.pk is None:
+                    if not field.remote_field.multiple:
+                        delattr(obj, field.related.get_cache_name())
```
Comment (timgraham) [inline django/db/models/base.py]: field.remote_field here too (you may squash your commits when updating, thanks)

File: tests/foreign_object/tests.py [hunk 4]
```diff
@@ -391,3 +391,10 @@ def test_batch_create_foreign_object(self):
         """ See: https://code.djangoproject.com/ticket/21566 """
         objs = [Person(name="abcd_%s" % i, person_country=self.usa) for i in range(0, 5)]
         Person.objects.bulk_create(objs, 10)
+
+    def test_error_unsaved_object(self):
```
Comment (timgraham) [inline tests/foreign_object/tests.py]: Does this test capture something that the tests in `many_to_one/tests.py` don't? It's a bit confusing to add it as part of this change as it doesn't appear to be related to the bug.

**Comment by raphaelmerx** [comment]
buildbot, test this please.

File: tests/foreign_object/tests.py (hunk 4 above)
Comment (raphaelmerx) [inline tests/foreign_object/tests.py]: I hadn't seen that this was handled in `many_to_one/tests.py`. Just removed that extra test.

**Comment by timgraham** [comment]
merged in f5a33e4840d3ad4d1199e99f5a17a9af1d2176f9, thanks!
